### ISSUE TITLE
word-count: leading apostrophe not tested

### ISSUE THREAD

**Issue by roadfoodr** [exercism/problem-specifications#1539]
The `word-count` test suite tests for inner apostrophes (should be kept) and surrounding apostrophes (should be stripped), but does not test for a contraction that involves only a leading apostrophe (should be kept).  I'd suggest:

```
    {
      "description": "with leading apostrophes",
      "property": "countWords",
      "input": {
        "sentence": "My country 'tis of thee"
      },
      "expected": {
        "my": 1,
        "country": 1,
        "'tis": 1,
        "of": 1,
        "thee": 1
      }
    },
```

We could also consider testing to keep trailing apostrophes in contracted words ("somethin' is better 'n nothin'"), but we would then also be keeping trailing apostrophes in plural possessives ("the pedants' arguments").  (Arguably those should be kept; after all, they are kept in singular possessives.)

If there is consensus around desired behavior here I can submit a pull request.  NOTE: this will likely break many existing solutions.

**Comment by rpottsoh** [comment]
I think this would be a good addition.  Broke my solution.

**Comment by yawpitch** [comment]
At the moment I'm rather on the fence on this one. It's a good point about a general-use word counter, but I'm not sure that expanding the definition of a _word_ to incorporate archaic contractions (ie "'twas", "'twasn't", "'tis") and informal contractions (ie "'nuthin", "something'") actually does anything to improve the pedagogical value of this exercise. Instead I'd say that all those expansions would do is substantially increase the complexity of the exercise without equally substantially increases the number of language features a student is exposed to.

Plus we're already accepting integers but ignoring _most_ other numerical forms (ie "10.0", "1e12", "6.02e10<sup>23</sup>", "$1K"). I'd argue that expanding the scope of the exercise to incorporate them would bring some useful new skills into play, but might also border on cruelty, if not outright sadism. IMHO it's a reasonable compromise to say that we're _not_ asking for a perfectly functioning general English vernacular word counter, but merely one that satisfies counting of a focussed subset of "words".

The plural possessive form is common enough in formal modern English that it might be worthy of a test case -- since it would introduce the notion of the positive look-ahead in regular expression-based solutions -- but since not all languages offer look-aheads and look-behinds I'm not entirely sure it's something that shouldn't be added on a track-by-track basis instead.

But I'd love to hear more discussion on it.

**Comment by F3PiX** [comment]
Thanks @roadfoodr for your thoughts! 
I also find that it's adding complexity without adding idiomatic use of the languages. It's without debate that it would make a better word counter, that's for sure, but we're not asking for the perfect word counter. (As @yawpitch pointed about with the numbers.) And we shouldn't. 
As the exercise has a low complexity, and therefor is a good candidate for tracks to have as one of the easier exercises, I'd vote against adding complexity.  

@roadfoodr For your information: my opinion is heavily colored by the Track Anatomy Project, where we're restructuring tracks with a well designed progression through the track. That's the reason why I treasure the non-complexity of exercises that are designed to be of low complexity. 

I'd vote for not adding this test.

**Comment by roadfoodr** [comment]
Thanks for the discussion so far.  As a recent first-time completer of this exercise, the current test that stood out to me as a difficulty-ramping edge case is the "surrounded by single quotes" test case.  (That made me start thinking about whether there are any situations where leading or trailing single quotes should be kept.)  It might be worth thinking about whether that edge case in particular meets the complexity/pedagogical targets being described here.

**Comment by F3PiX** [comment]
I totally agree, I don't feel the apostroph test is adding enough added value to justify it.

**Comment by bencoman** [comment]
For non-native english speakers we should keep away from archaic forms.  A better choice would be a contraction of "until" - "He waited 'til Joe arrived."  The trailing ownership-apostrophe is also common enough to include. 

I do agree that the focus should be away from perfecting the domain and providing just-enough-complexity to keep students engaged while learning language features.  However as long as spaces are not being included between single-quotes** I don't think this adds much complexity. Regular expression look-ahead won't be required.  The tokenization valid-character-list just expands to alpha+num+quote.  The resultant tokens then have the quotes stripped out to get words for comparison.  

Actually independently I was considering that a more practical domain solution would be to ignore quotations.  Like this...
{
      "description": "inconsistent punctuation",
      "property": "countWords",
      "input": {
        "sentence": "Joe can't spell 'cant' consistently"
      },
      "expected": {
        "joe": 1,
        "cant": 2,
        "spell": 2,
        "consistently": 1
      }

Consistently stripping quotes rather than sometimes (e.g. to deal with "can't") would simplify the exercise.  Would that make it too simple?   Students might just strip single-quotes from the input string before further processing. 

After all that, if @F3PiX has experience from an overall analysis of exercises, I'd go with his opinion of what the complexity of this exercise should be.

** The problem description should be explicit about this simplification, rather than just implicit from the tests.

**Comment by SleeplessByte** [comment]
I actually think that it makes sense to strip _all_ characters from Word Count and add Word Count II or something like that as an exercise _with_ numbers _and_ apostrophe's etc. cc @F3PiX

**Comment by yawpitch** [comment]
@bencoman side note, but "'til" and "'till" aren't actually contractions of "until", they're modern bastardizations of "till" that most likely happen because people have seen the archaic "'tis" and think it applies. And lookahead's would be required for properly handling the plural possessive because it _must_ be an "s" followed by an apostrophe to be a plural possessive. Lots and lots of solutions I've seen will consider "1'''''''''''''''''''''''''''''''''''p" a word when it clearly isn't.

@roadfoodr there isn't exactly a single-quotes surround case, there's an explicit case for a single quoted word and a second case that incidentally handles the double quoted one. It would probably be best to update that test so that both quote forms were included, ie "Joe can't tell between 'large' and \"large\"." to avoid confusion.

This might very well be my interpretation, but for most languages that implement this problem it looks like what's desired is the ability to solve it with _either_ common string methods or regular expressions / pattern matching, with a bit of a preferences towards regular expressions if available.

The current _implict_ definition provides that; a _word_ for the purposes of `word-count` is:

1. an integer composed of one or more ASCII digits (ie 0-9, as in "1234").
2. a word composed of one or more ASCII letters (ie a-z, as in "they") _optionally_ followed by an apostrophe AND one or more ASCII letters (as simple contraction as in "they're").

IMHO that's a good enough approximation; it leaves the door open for solving via string methods with just splitting and stripping, but it's got a considerably better solution using regex to search for non-overlapping matches to a pattern that can be described by `[0-9]+|[a-z]+('[a-z]+)?` or the language-local equivalent.

I'd say we'd get far enough by making that definition explicit in the Instructions/README, and possibly adding something like "To keep this simple you do not need to consider the plural possessive form (as in "the ants' nest"), but it would make a good bonus exercise." or something similar, for those who want to have a stretch goal.

We _could_ also explicitly state that archaic and informal contractions aren't included, but personally I think that just muddies the waters, especially for non-English speakers.

**Comment by SleeplessByte** [comment]
> We could also explicitly state that archaic and informal contractions aren't included, but personally I think that just muddies the waters, especially for non-English speakers.

Agreed with this!

**Comment by ErikSchierboom** [comment]
> I actually think that it makes sense to strip all non letter characters from Word Count and add Word Count II or something like that as an exercise with numbers and apostrophe's etc. cc @F3PiX

I like this suggestion!

**Comment by SleeplessByte** [comment]
Maud is currently OOO I believe, but I think she'd be in favour of _not_ making this exercise harder and instead add a more complicated version II.

@roadfoodr @yawpitch are you interested in drafting that up?

**Comment by yawpitch** [comment]
If we've reached rough consensus on leaving this exercise at its current level of complexity I'm happy to add some wording as described above to the README and tests to that end and I'll submit a PR.

I'm not _certain_ we've reached that consensus, as there's at least two votes for _reducing_ the complexity of this exercise and removing the contractile apostrophe and surrounding quotes (and possibly numbers, that's a bit unclear) and moving all of that to a proposed `word-count-2`.

To be perfectly clear I'm in favour of leaving this exercise at its current level of complexity, making no change other than to update the README to remove ambiguity around what a "word" is for the scope of _this_ exercise.

**Comment by SleeplessByte** [comment]
I'm only in favour of _reducing_ it here _if_ we create `Word Count II`.

**Comment by yawpitch** [comment]
> I'm only in favour of _reducing_ it here _if_ we create `Word Count II`.

Ok, that's good to know. Personally I'm biased against any change that either breaks existing valid solutions without adding _obvious_ pedagogic value to an exercise, but also any change that invalidates the time investment students have already made in their solution by _reducing_ the existing value of the exercise. Splitting the learning opportunities over multiple versions of the same exercise, IMHO, only makes sense if there's some discreet set of skills / tags that are mutually exclusive to each ... so if `word-count` is aimed at basic string methods (replace, split, trim, etc) and `word-count-2`  is pattern matching and regex, then such a split might make some sense to me.

That said I feel like `word-count` is currently positioned in Python, for instance, after _several_ exercises that can be completed with string methods, and is well positioned to be the first real opportunity for an introduction to regex ... but only _because_ of the contractile apostrophe.

So for that track at least I'd tend to find myself voting to eliminate a _reduced_ `word-count` entirely from **Core** (if not from the Track) in favour of the `word-count-2` that received that reduction. Which is why I'm not sure about "robbing from Peter to pay Paul" here; I'd rather leave `word-count` as is and add a separate, perhaps differently-focused exercise (ie `find-contractions`, or equivalent) instead of revisiting word counting.

**Comment by ErikSchierboom** [comment]
> I'm only in favour of reducing it here if we create Word Count II.

Agreed.

**Comment by yawpitch** [comment]
Just pushed a PR #1542 what seeks to address improving the description of `word-count` with an eye towards maintaining the current level of complexity. I propose we _do not_ create a Word Count II at this point, as I think there are better exercises for introducing the few new concepts required to handle the idiosyncrasies of apostrophe use in English.

**Comment by SleeplessByte** [comment]
I think that PR makes a lot of sense @yawpitch . Much better like that.

**Comment by ErikSchierboom** [comment]
Yep, great PR!

**Comment by yawpitch** [comment]
Closed by merge of #1542, tabling discussion of **Word Count II**. If anyone wants to open an issue / PR for that they can reference this issue.
